README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# surveydataset

<!-- badges: start -->

[![lifecycle](https://lifecycle.r-lib.org/articles/figures/lifecycle-experimental.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.7013164.svg)](https://zenodo.org/record/7013164#.YwFmc3ZBzIU)
[![devel-version](https://img.shields.io/badge/devel%20version-0.1.0-blue.svg)](https://github.com/dataobservatory-eu/surveydataset)
[![dataobservatory](https://img.shields.io/badge/ecosystem-dataobservatory.eu-3EA135.svg)](https://dataobservatory.eu/)
[![Follow
rOpenGov](https://img.shields.io/twitter/follow/ropengov.svg?style=social)](https://twitter.com/intent/follow?screen_name=ropengov)
[![Follow
author](https://img.shields.io/twitter/follow/digitalmusicobs.svg?style=social)](https://twitter.com/intent/follow?screen_name=digitalmusicobs)
[![Codecov test
coverage](https://codecov.io/gh/dataobservatory-eu/surveydataset/branch/master/graph/badge.svg)](https://app.codecov.io/gh/dataobservatory-eu/surveydataset?branch=master)
[![R-CMD-check](https://github.com/dataobservatory-eu/surveydataset/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/dataobservatory-eu/surveydataset/actions/workflows/R-CMD-check.yaml)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/dataobservatory-eu/surveydataset?branch=master&svg=true)](https://ci.appveyor.com/project/dataobservatory-eu/surveydataset)
<!-- badges: end -->

The goal of the ‘surveydataset’ package is to create special datasets
for social sciences surveys that are easier to release, exchange and
reuse. It builds on the [dataset](http://dataset.dataobservatory.eu/)
package which creates data frames that have clear structural and
referncial metadata to be used in statistical exchanges or placed in
scientific repositories. The ‘surveydataset’ package extends this
functionality with the pecularities of social sciences surveys, and
offers a data frame like container to be used with
[DDIwR](https://CRAN.R-project.org/package=DDIwR) for survey
documentation and
[retroharmonize](https://retroharmonize.dataobservatory.eu/) for the
retrospective harmonization of survey datasets.

## Installation

You can install the development version of surveydataset like so:

``` r
remotes::install_github('dataobservatory-eu/surveydataset')
```

## Read in a file

``` r
library(surveydataset)
require(haven)
path <- system.file("ZA5688_v6_sample.sav", package = "surveydataset")
survey_raw       <- read_spss_declared(path)
```

We have created a wrapper around
[haven::read_spss()](https://haven.tidyverse.org/reference/read_spss.html)
and `declared::as_declared()` to read in surveys in a less ambigous
format:

``` r
summary(as.factor(survey_raw$d25))
#> Rural area or village     Small/middle town            Large town 
#>                    34                    41                    25
summary(survey_raw$d25)
#>    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#>    1.00    1.00    2.00    1.91    2.25    3.00
```

The `read_spss_declared()` imports SPSS survey data into a
[tibble](https://CRAN.R-project.org/package=tible), and changes the
[labelled](https://CRAN.R-project.org/package=labelled) vectors
(variable columns) to
[declared](https://CRAN.R-project.org/package=declared) columns.

## Remove constans

``` r
small_survey <- survey_raw[, c(1:6, 7, 10:14, 33)] # Only to keep example easier to read
head(small_survey)
#> # A tibble: 6 × 13
#>           studyno1       studyno2 doi   version     edition        survey caseid
#>          <int+lbl>      <int+lbl> <chr> <chr>     <int+lbl>     <int+lbl>  <dbl>
#> 1 5688 [GESIS STU… 5688 [GESIS S… doi:… 6.0.0 … 2 [ARCHIVE… 792 [Eurobar… 752670
#> 2 5688 [GESIS STU… 5688 [GESIS S… doi:… 6.0.0 … 2 [ARCHIVE… 792 [Eurobar… 360001
#> 3 5688 [GESIS STU… 5688 [GESIS S… doi:… 6.0.0 … 2 [ARCHIVE… 792 [Eurobar…   3507
#> 4 5688 [GESIS STU… 5688 [GESIS S… doi:… 6.0.0 … 2 [ARCHIVE… 792 [Eurobar…    876
#> 5 5688 [GESIS STU… 5688 [GESIS S… doi:… 6.0.0 … 2 [ARCHIVE… 792 [Eurobar…    403
#> 6 5688 [GESIS STU… 5688 [GESIS S… doi:… 6.0.0 … 2 [ARCHIVE… 792 [Eurobar…   1759
#> # … with 6 more variables: isocntry <chr>, wextra <dbl>, qb1_1 <int+lbl>,
#> #   qb1_2 <int+lbl>, qb1_3 <int+lbl>, d25 <int+lbl>
#> # ℹ Use `colnames()` to see all variable names
```

``` r
my_survey <- survey(
  small_survey,
  Dimensions = c("isocntry", "d25"),
  Measures = c("qb1_1","qb1_2", "qb1_3"), 
  Attributes = c("wextra", "caseid"), 
  Constants = c(1:5),
  Identifier =  small_survey$doi[1], 
  Label = as.character(as.factor(small_survey$survey[1])),
  Title = "Eurobarometer 79.2 (Small Demo Sample)", 
  Publisher = "GESIS", 
  Issued = substr("6.0.0 (2016-08-05)", 8, 17),
  Creator = person("GESIS"))

dataset::version(my_survey) <- small_survey$version[1]
```

``` r
summary(my_survey)
#>      survey        caseid         isocntry             wextra       
#>  Min.   :792   Min.   :    18   Length:100         Min.   :  305.7  
#>  1st Qu.:792   1st Qu.:  1287   Class :character   1st Qu.: 2955.1  
#>  Median :792   Median :181058   Mode  :character   Median : 6688.8  
#>  Mean   :792   Mean   :263517                      Mean   :13991.8  
#>  3rd Qu.:792   3rd Qu.:360175                      3rd Qu.:14649.0  
#>  Max.   :792   Max.   :988853                      Max.   :79029.3  
#>                                                                     
#>      qb1_1           qb1_2          qb1_3           d25      
#>  Min.   :1.000   Min.   :1.00   Min.   :1.00   Min.   :1.00  
#>  1st Qu.:1.000   1st Qu.:1.00   1st Qu.:1.00   1st Qu.:1.00  
#>  Median :1.000   Median :1.00   Median :1.00   Median :2.00  
#>  Mean   :1.293   Mean   :1.65   Mean   :1.47   Mean   :1.91  
#>  3rd Qu.:1.000   3rd Qu.:2.00   3rd Qu.:2.00   3rd Qu.:2.25  
#>  Max.   :4.000   Max.   :4.00   Max.   :4.00   Max.   :3.00  
#>  NA's   :1
```

``` r
my_attributes <- attributes(my_survey)
my_attributes$names <- my_attributes$row.names <- NULL; 
my_attributes
#> $dimensions
#>             names            class
#> isocntry isocntry        character
#> d25           d25 declared|integer
#>                                                                                                                                                     isDefinedBy
#> isocntry https://purl.org/linked-data/cube|https://raw.githubusercontent.com/UKGovLD/publishing-statistical-data/master/specs/src/main/vocab/sdmx-attribute.ttl
#> d25      https://purl.org/linked-data/cube|https://raw.githubusercontent.com/UKGovLD/publishing-statistical-data/master/specs/src/main/vocab/sdmx-attribute.ttl
#>                 codeList
#> isocntry not yet defined
#> d25      not yet defined
#> 
#> $measures
#>       names            class                       isDefinedBy       codeListe
#> qb1_1 qb1_1 declared|integer https://purl.org/linked-data/cube not yet defined
#> qb1_2 qb1_2 declared|integer https://purl.org/linked-data/cube not yet defined
#> qb1_3 qb1_3 declared|integer https://purl.org/linked-data/cube not yet defined
#> 
#> $attributes
#>         names   class
#> caseid caseid numeric
#> wextra wextra numeric
#>                                                                                                                                                   isDefinedBy
#> caseid https://purl.org/linked-data/cube|https://raw.githubusercontent.com/UKGovLD/publishing-statistical-data/master/specs/src/main/vocab/sdmx-attribute.ttl
#> wextra https://purl.org/linked-data/cube|https://raw.githubusercontent.com/UKGovLD/publishing-statistical-data/master/specs/src/main/vocab/sdmx-attribute.ttl
#>              codeListe
#> caseid not yet defined
#> wextra not yet defined
#> 
#> $Type
#>       resourceType resourceTypeGeneral
#> 1 DCMITYPE:Dataset             Dataset
#> 
#> $Title
#>                                    Title titleType
#> 1 Eurobarometer 79.2 (Small Demo Sample)     Title
#> 
#> $Identifier
#> [1] "doi:10.4232/1.12577"
#> 
#> $Creator
#> [1] "GESIS"
#> 
#> $Source
#> [1] NA
#> 
#> $Publisher
#> [1] NA
#> 
#> $Rights
#> [1] NA
#> 
#> $Description
#> [1] NA
#> 
#> $Size
#> [1] "33.08 kB [32.3 KiB]"
#> 
#> $Date
#> [1] "2022-08-22"
#> 
#> $datasetAttributes
#>   attribute                 value
#> 1  studyno1 GESIS STUDY ID ZA5688
#> 2  studyno2 GESIS STUDY ID ZA5688
#> 3       doi   doi:10.4232/1.12577
#> 4   version    6.0.0 (2016-08-05)
#> 5   edition       ARCHIVE EDITION
#> 
#> $class
#> [1] "dataset"    "data.frame"
#> 
#> $Version
#> [1] "6.0.0 (2016-08-05)"
```

``` r
utils::toBibtex(dataset::bibentry_dataset(my_survey))
```

## Further work

The `surveydataset` package is in a very early stage of its development.
For citations please refer to the current development version as:

``` r
citation('surveydataset')
#> 
#> To cite surveydataset in publications use:
#> 
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Misc{,
#>     year = {2022},
#>     publisher = {Reprex},
#>     title = {surveydataset: An R Package For Findable, Accessible, Interoperable, and Reusable Survey Datasets},
#>     author = {Daniel Antal},
#>     url = {https:://surveydataset.dataobservatory.eu/},
#>     version = {0.1.0},
#>     doi = {10.5281/zenodo.7013164},
#>   }
```

## Code of Conduct

Please note that the surveydataset project is released with a
[Contributor Code of
Conduct](https://contributor-covenant.org/version/2/1/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# surveydataset 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(surveydataset)

test_check("surveydataset")


tests/testthat/test-survey.R
path <- system.file("ZA5688_v6_sample.sav", package = "surveydataset")
survey_raw       <- read_spss_declared(path)


Identifier <- survey_raw$doi[1]
version_info <- survey_raw$version[1]

y <- survey(
  x = survey_raw,
  Dimensions = c(10, 22:33),
  Measures = c(13:21),
  Attributes = c(7:9, 11),
  Constants = c(1:6),
  Title = "Test Survey",
  Identifier = survey_raw$doi[1],
  Label = as.character(as.factor(survey_raw$survey[1]))
)


test_that("survey() works", {
  expect_equal(ncol(y), 27)
})


tests/testthat/test-read_spss_declared.R
haven_tested <- haven::read_sav(file, encoding = NULL, user_na = FALSE, col_select = NULL, skip = skip, n_max = n_max,
                .name_repair = .name_repair)

tested <- read_sav_declared(system.file("ZA5688_v6_sample.sav", package = "surveydataset"),
                  encoding = NULL, user_na = FALSE, col_select = NULL, skip = skip, n_max = n_max,
                  .name_repair = .name_repair)

test_that("read_sav_declared() works", {
  expect_equal(nrow(tested), nrow(haven_tested))
  expect_equal(ncol(haven_tested), ncol(tested))
})

tested_spss <- read_spss_declared(
  file = system.file("ZA5688_v6_sample.sav", package = "surveydataset"),
  encoding = NULL, user_na = FALSE, col_select = NULL, skip = skip, n_max = n_max,
  .name_repair = .name_repair)


test_that("read_spss_declared() works", {
  expect_equal(nrow(tested), nrow(haven_tested))
  expect_equal(ncol(haven_tested), ncol(tested))
})


R/read_spss_declared.R
#' @title Read and write SPSS files
#' @description
#' `read_sav()` reads both `.sav` and `.zsav` files.
#' `read_por_declared()` reads `.por` files.
#' `read_spss_declared()` uses either `read_por_declared()` or `read_sav_declared()` based on the
#' file extension.
#' @details
#' Currently haven can read and write logical, integer, numeric, character
#' and factors. This is not ideal for social science surveys, and therefore we created a
#' wrapper haven's functions to read the data into declared types.
#' @importFrom haven read_spss read_por read_sav
#' @param file The file to read.
#' @param col_select One or more selection expressions, like in
#'   [dplyr::select()]. Use `c()` or `list()` to use more than one expression.
#'   See `?dplyr::select` for details on available selection options. Only the
#'   specified columns will be read from `data_file`.
#' @param skip Number of lines to skip before reading data.
#' @param n_max Maximum number of lines to read.
#' @param encoding The character encoding used for the file. The default,
#'   `NULL`, use the encoding specified in the file, but sometimes this
#'   value is incorrect and it is useful to be able to override it.
#' @return A tibble, data frame variant with nice defaults.
#'
#'   Variable labels are stored in the "label" attribute of each variable.
#'   It is not printed on the console, but the RStudio viewer will show it.
#' @name read_spss_declared
#' @examples
#' path <- system.file("examples", "ZA5688_v6_sample.sav", package = "surveydataset")
#' read_sav_declared(path)
#'
#' tmp <- tempfile(fileext = ".sav")
#' read_sav_declared(tmp)
NULL

#' @export
#' @rdname read_spss_declared
read_sav_declared <- function(file,
                              encoding = NULL,
                              user_na = FALSE,
                              col_select = NULL,
                              skip = 0,
                              n_max = Inf,
                              .name_repair = "unique") {

  argg <- list(encoding = encoding,
               col_select = col_select)

  if (is.null(col_select)) {
    tmp <- haven::read_sav(file, encoding = argg$encoding, user_na = user_na, skip = skip, n_max = n_max, .name_repair = .name_repair)

  } else {
    tmp <- haven::read_sav(file, encoding = argg$encoding,
                           user_na = user_na,
                           col_select = col_select,
                           skip = skip, n_max = n_max,
                           .name_repair = .name_repair)
  }

  declared::as.declared(tmp)
}

#' @export
#' @rdname read_spss_declared
read_por_declared <- function(file, user_na = FALSE, col_select = NULL, skip = 0, n_max = Inf, .name_repair = "unique") {

  argg <- list(encoding = encoding,
               col_select = col_select)

  if (is.null(col_select)) {
    tmp <- haven::read_por(file, encoding = argg$encoding, user_na = user_na, skip = skip, n_max = n_max, .name_repair = .name_repair)

  } else {
    tmp <- haven::read_por(file, encoding = argg$encoding,
                           user_na = user_na,
                           col_select = col_select,
                           skip = skip, n_max = n_max,
                           .name_repair = .name_repair)
  }
  declared::as.declared(tmp)
}


#' @export
#' @rdname read_spss_declared
#' @param user_na If `TRUE` variables with user defined missing will
#'   be read into [labelled_spss()] objects. If `FALSE`, the
#'   default, user-defined missings will be converted to `NA`.
read_spss_declared <- function(file,
                               user_na = FALSE,
                               col_select = NULL,
                               skip = 0,
                               n_max = Inf,
                               .name_repair = "unique") {


  if (is.null(col_select)) {
    tmp <- haven::read_spss(file, user_na = user_na, skip = skip, n_max = n_max, .name_repair = .name_repair)

  } else {
    tmp <- haven::read_spss(file,
                            user_na = user_na,
                            col_select = col_select,
                            skip = skip, n_max = n_max,
                            .name_repair = .name_repair)
  }
  declared::as.declared(tmp)
}



R/constant_to_attributes.R
#' @title Constants to attributes
#' @description Move columns that have only a single value present the dataset into the
#' attributes of a dataset.
#' @param x A data frame or similar R object.
#' @param constant_colums A selection of constant columns. Defaults to \code{'all'}.
#' @param attribute_name The name of the attribute which will containt the \code{constant_columns}.
#' @return A subsetted data frame, which includes the unique values of the constant columns
#' among the attributes of the R object.
#' @family dimension reduction functions
#' @examples
#' path <- system.file("ZA5688_v6_sample.sav", package = "surveydataset")
#' survey_raw <- read_spss_declared(path)
#' survey_raw <- survey_raw[, c(1:5, 7, 10:12 )]
#' constant_to_attributes(x = survey_raw, constant_columns = 1:5)
#'
#' my_survey <- constant_to_attributes(x = survey_raw, constant_columns = 1:5)
#' attr(my_survey, "datasetAttributes")
#' @export

constant_to_attributes <- function(x,
                                   constant_columns,
                                   attribute_name = "datasetAttributes") {

  if(is.numeric(constant_columns)) {

    if ( !all(constant_columns %in% seq_along(x))) {

      message("constant_to_attributes(x, constant_columns, ...) - not in  x:",
              paste(constant_columns[!constant_columns %in% seq_along(x)], collapse = ', ')
              )
      constant_columns <- constant_columns[constant_columns %in% seq_along(x)]
    }
    constant_columns <- names(x)[constant_columns]
  }

  columns_not_present <- not_in_df(x, constant_columns)

  available_columns <- constant_columns[! constant_columns %in% columns_not_present ]

  subsetted_df <- subset(x, select = available_columns)

  not_unique_columns <- subsetted_df[, ! vapply(subsetted_df, unique_columns, logical(1))]

  if (ncol(not_unique_columns)>0) {
    message("constant_to_attributes(x, constant_columns) - not constant: ", paste(names(not_unique_columns), collapse = (', ')))
    subsetted_df <- subsetted_df[, which(! names(subsetted_df) %in% names(not_unique_columns))]
  }

  dataset_attributes <- data.frame(
    attribute = names(subsetted_df),
    value = as.character(vapply(subsetted_df, function(x) as.character(unique(x)), character(1)))
  )

  attr(x, which=attribute_name) <- dataset_attributes

  x[, names(x) %in% names(subsetted_df)] <- NULL

  x
}


#' @keywords internal
not_in_df <- function(x, columns) {
  names(x)[which (!columns %in% names(x))]
}

#' @keywords internal
unique_columns <- function(x) {
  length(unique(x))==1
}


R/survey.R
#' @title Create a survey
#' @description Create a dataset suitable for working with social sciences surveyd data.
#' @importFrom dataset dataset
#' @inheritParams dataset::dataset
#' @return A well-described and formatted data frame with survey data and rich metadata.
#' @examples
#' path <- system.file("ZA5688_v6_sample.sav", package = "surveydataset")
#' survey_raw       <- read_spss_declared(path)
#'
#' Identifier <- survey_raw$doi[1]
#'
#' y <- survey(
#'   x = survey_raw,
#'   Dimensions = c(10, 22:33),
#'   Measures = c(13:21),
#'   Attributes = c(7:9, 11),
#'   Constants = c(1:6),
#'   Title = "Test Survey",
#'   Identifier = survey_raw$doi[1],
#'   Label = as.character(as.factor(survey_raw$survey[1]))
#' )
#' @export

survey <- function(x,
                   Dimensions, Measures, Attributes, Constants,
                   Title = NULL, Identifier,
                   Label = NULL, Creator = NULL,  Publisher = NULL, Issued = NULL,
                   Subject = NULL ) {

  tmp <- dataset(x,
                 Dimensions=Dimensions,
                 Measures=Measures,
                 Attributes=Attributes,
                 Title = Title,
                 Label = Label,
                 Creator = Creator,
                 Identifier = Identifier,
                 Subject = Subject)


  tmp <- constant_to_attributes(x=tmp, constant_columns = Constants)

  tmp
}

